### PR TITLE
Feature/batch

### PR DESCRIPTION
**Pull request by AlexB52**
Ready to be reviewed

### CODE REVIEW

File: app/controllers/admin/batches_controller.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
+class Admin::BatchesController < ApplicationController
+  before_action :set_admin_batch, only: [:show, :edit, :update, :destroy]
+
+  def new
+    @batch = Batch.new
+    @roast = Roast.find(params[:roast])
+    @batch.roast = Roast.find(params[:roast])
```
Comment (AlexB52) [inline app/controllers/admin/batches_controller.rb]: We can reuse the `@roast` fetched and assign it directly to the `@batch.roast`. 
By calling `Roast.find` twice we are querying the database twice for the same data.
```ruby
@batch.roast = @roast
```

File: app/controllers/admin/batches_controller.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      redirect_to admin_roast_path(@batch.roast_id), notice: 'Batch was successfully updated.'
+    else
+      render :edit
+    end
+  end
+
+  def destroy
+    roast = @batch.roast_id
+    @batch.destroy
+    redirect_to admin_roast_url(roast), notice: 'Batch was successfully destroyed.'
```
Comment (AlexB52) [inline app/controllers/admin/batches_controller.rb]: Naming glitch here between what we assign and how we name it.
```
roast = @batch.roast
OR
roast_id = @batch.roast_id
```

File: app/controllers/admin/batches_controller.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  def update
+    if @batch.update(admin_batch_params)
+      redirect_to admin_roast_path(@batch.roast_id), notice: 'Batch was successfully updated.'
+    else
+      render :edit
+    end
+  end
+
+  def destroy
+    roast = @batch.roast_id
```
Comment (AlexB52) [inline app/controllers/admin/batches_controller.rb]: Naming glitch here between what we assign and how we name it.
```
roast = @batch.roast
OR
roast_id = @batch.roast_id
```

File: app/controllers/admin/batches_controller.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    redirect_to admin_roast_url(roast), notice: 'Batch was successfully destroyed.'
+  end
+
+  private
+  # Use callbacks to share common setup or constraints between actions.
+  def set_admin_batch
+    @batch = Batch.find(params[:id])
+  end
+
+
```
Comment (AlexB52) [inline app/controllers/admin/batches_controller.rb]: There are some code styles that are seen as best practices.
It is quite extensive to read but if you follow this guide you are good to go on most ruby projects. https://github.com/bbatsov/ruby-style-guide

File: app/controllers/admin/batches_controller.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  # Never trust parameters from the scary internet, only allow the white list through.
+  def admin_batch_params
+    params.require(:batch).permit(:roast_id, :start_date, :cost, :amount_purchased)
+  end
+
+  def get_all_roasts
+    @roasts = Roast.all
+  end
+
```
Comment (mattlee68) [inline app/controllers/admin/batches_controller.rb]: Remove break line

File: app/controllers/admin/roasts_controller.rb [hunk 6]
```diff
@@ -1,10 +1,10 @@
-class Admin::RoastsController < ApplicationController
-  http_basic_authenticate_with name: ENV['ADMIN_USERNAME'], password: ENV['ADMIN_PASSWORD']
+class Admin::RoastsController < Admin::ApplicationController
+
```
Comment (mattlee68) [inline app/controllers/admin/roasts_controller.rb]: Remove break line

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -1,2 +1,6 @@
 module ApplicationHelper
+
```
Comment (mattlee68) [inline app/helpers/application_helper.rb]: Remove break line

File: app/controllers/admin/roasts_controller.rb [hunk 8]
```diff
@@ -19,7 +19,6 @@ def edit
   # POST /admin/roasts
   def create
     @roast = Roast.new(admin_roast_params)
-
```
Comment (mattlee68) [inline app/controllers/admin/roasts_controller.rb:22]: Add break line

File: app/views/admin/roasts/_roast.html.erb [hunk 9]
```diff
@@ -0,0 +1,29 @@
+
```
Comment (mattlee68) [inline app/views/admin/roasts/_roast.html.erb]: Remove break line

File: app/views/admin/roasts/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+      <td><%= batch_presenter.cost_per_gram %></td>
+      <td><%= batch_presenter.amount_purchased %></td>
+      <td><%= link_to 'Edit', edit_admin_batch_path(batch_presenter) %></td>
+      <td><%= link_to 'Delete', admin_batch_path(batch_presenter), method: :delete, data: { confirm: 'Are you sure you want to delete this batch?' } %></td>
+    </tr>
+  <% end %>
+</table>
+<% end %>
+
+
```
Comment (mattlee68) [inline app/views/admin/roasts/show.html.erb]: Remove break line

File: test/controllers/admin/batch_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,66 @@
+require 'test_helper'
+
+class Admin::RoastsControllerTest < ActionDispatch::IntegrationTest
+
```
Comment (mattlee68) [inline test/controllers/admin/batch_controller_test.rb]: remove break line

File: test/controllers/admin/batch_controller_test.rb [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  setup do
+    @header = {"Authorization" => ActionController::HttpAuthentication::Basic.encode_credentials(ENV["ADMIN_USERNAME"], ENV["ADMIN_PASSWORD"])}
+
```
Comment (mattlee68) [inline test/controllers/admin/batch_controller_test.rb]: remove break line

File: test/controllers/admin/batch_controller_test.rb [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+  end
+
+  test "batch authentication requests" do
+    check_for_authentication([
+      Request.new(:get, new_admin_roast_url),
+      Request.new(:get, edit_admin_batch_url(@batch)),
+      Request.new(:post, admin_batches_url),
+      Request.new(:delete, admin_batch_url(@batch)),
+      Request.new(:patch, admin_batch_url(@batch))
+
```
Comment (mattlee68) [inline test/controllers/admin/batch_controller_test.rb]: remove break line

File: test/controllers/admin/batch_controller_test.rb [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+  test "should get new batch" do
+    get new_admin_batch_path(:roast => @roast), headers: @header
+    assert_response :success
+  end
+
+  test "should get new batch - with no roast" do
+    get new_admin_batch_path, headers: @header
+    assert_response :success
+
```
Comment (mattlee68) [inline test/controllers/admin/batch_controller_test.rb]: remove break line

File: test/controllers/admin/batch_controller_test.rb [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    assert_redirected_to admin_roast_url(@batch.roast)
+  end
+
+  test "should destroy batch" do
+    assert_difference('Batch.count', -1) do
+      delete admin_batch_url(@batch), headers: @header
+    end
+    assert_redirected_to admin_root_url
+  end
+
```
Comment (mattlee68) [inline test/controllers/admin/batch_controller_test.rb]: remove these two break lines

File: app/helpers/application_helper.rb [hunk 16]
```diff
@@ -1,2 +1,5 @@
 module ApplicationHelper
+  def format_roast_options(roasts)
```
Comment (mattlee68) [inline app/helpers/application_helper.rb:2]: Move this to roast_helper.rb

File: app/controllers/admin/welcome_controller.rb [hunk 17]
```diff
@@ -0,0 +1,6 @@
+class Admin::WelcomeController < Admin::ApplicationController
+  def index
+    @roasts = Roast.all
```
Comment (marcusbaguley) [inline app/controllers/admin/welcome_controller.rb:3]: Not sure these will be needed but understand you have them just for debug purposes...

**Review by marcusbaguley — APPROVED** [review]
👍

**Pull request merged**